src/compiler/core.ts

```diff
@@ -2159,8 +2174,19 @@ namespace ts {
         return (arg: T) => f(arg) && g(arg);
     }
 
-    export function or<T>(f: (arg: T) => boolean, g: (arg: T) => boolean): (arg: T) => boolean {
-        return arg => f(arg) || g(arg);
+    export function or<T extends unknown>(...fs: ((arg: T) => boolean)[]): (arg: T) => boolean {
+        return arg => {
+            for (const f of fs) {

```

Comment:
How do you choose between for-of, for-increment, and `forEach`?

Reply:
My personal tendencies:

- for-of: lets you early return enclosing function, doesn’t give you element index
- for-increment: generally only use when element index is important, or if I need to start somewhere other than the first element
- `forEach`: gives you early termination of iteration but not return of the enclosing function, result transformation, and element index. It’s generally replaceable by for-of but is more elegant if you return something, like

```ts
const firstMatchingSignature = forEach(types, type => {
  return firstDefined(type.getCallSignatures(), somePredicateThing);
});
```

- `Array.prototype.forEach`: no early termination, no return value 👎

Reply:
`arg => forEach(fs, f => f(arg)) || false`?

Reply:
Yeah, that would work. I think I also have a tendency to implement abstractions with building blocks that are _less_ abstract than the function I’m implementing, if that makes sense. `or` and `forEach` are both fairly low-level atomic abstractions, so it makes sense to me that neither would depend on the other, but would rather be built with primitive loops. (I’ve never been consciously aware of this thought process; it’s typically just driven by a vague intuition.)

Reply:
I guess, in my mind, if `or` can straightforwardly be reduced to `forEach` it is higher-level.  `forEach` also seems like the sort of thing that might eventually be performance-tuned and it would be nice to pick up the benefits automatically.  I don't feel strongly about it though.